lib/action_subscriber/publisher/async/in_memory_adapter.rb

```diff
@@ -0,0 +1,153 @@
+
+module ActionSubscriber
+  module Publisher
+    module Async
+      class InMemoryAdapter
+        include ::ActionSubscriber::Logging
+
+        attr_reader :async_queue
+
+        def initialize
+          logger.info "Starting in-memory publisher adapter."
+
+          @async_queue = AsyncQueue.new
+        end
+
+        def publish(route, payload, exchange_name, options = {})
+          message = Message.new(route, payload, exchange_name, options)
+          async_queue.push(message)
+          nil
+        end
+
+        def shutdown!
+          max_wait_time = ::ActionSubscriber.configuration.seconds_to_wait_for_graceful_shutdown
+          started_shutting_down_at = ::Time.now
+
+          logger.info "Draining async publisher in-memory adapter queue before shutdown. Current queue size: #{async_queue.size}."
+          while async_queue.size > 0
+            if (::Time.now - started_shutting_down_at) > max_wait_time
+              logger.info "Forcing async publisher adapter shutdown because graceful shutdown period of #{max_wait_time} seconds was exceeded. Current queue size: #{async_queue.size}."
+              break
+            end
+
+            sleep 0.1
+          end
+        end
+
+        class Message
+          attr_reader :route, :payload, :exchange_name, :options
+
+          def initialize(route, payload, exchange_name, options)
+            @route = route
+            @payload = payload
+            @exchange_name = exchange_name
+            @options = options
+          end
+        end
+
+        class UnableToPersistMessageError < ::StandardError
+        end
+
+        class AsyncQueue
+          include ::ActionSubscriber::Logging
+
+          attr_reader :consumer, :queue, :supervisor
+
+          if ::RUBY_PLATFORM == "java"
+            NETWORK_ERRORS = [::MarchHare::Exception, ::Java::ComRabbitmqClient::AlreadyClosedException, ::Java::JavaIo::IOException].freeze
+          else
+            NETWORK_ERRORS = [::Bunny::Exception, ::Timeout::Error, ::IOError].freeze
+          end
+
+          def initialize
+            @queue = ::Queue.new
+            create_and_supervise_consumer!
+          end
+
+          def push(message)
+            # Default of 1_000_000 messages.
+            if queue.size > ::ActionSubscriber.configuration.async_publisher_max_queue_size
+              # Drop Messages if the queue is full and we were configured to do so.
+              return if ::ActionSubscriber.configuration.async_publisher_drop_messages_when_queue_full
+
+              # By default we will raise an error to push the responsibility onto the caller.
+              fail UnableToPersistMessageError, "Queue is full, messages will be dropped."
+            end
+
+            queue.push(message)
+          end
+
+          def size
+            queue.size
+          end
+
+        private
+
+          def await_network_reconnect
+            sleep ::ActionSubscriber::RabbitConnection::NETWORK_RECOVERY_INTERVAL
```

Comment:
I'm okay leaving this as a constant because if we end up using any async-adapter that goes to redis we probably won't be in charge of configuring the network recovery intervals inside the publishing process.